src/server/stream_family.cc

```diff
@@ -1640,6 +1850,54 @@ void StreamFamily::XRead(CmdArgList args, ConnectionContext* cntx) {
   }
 }
 
+void XReadGeneric(CmdArgList args, bool read_group, ConnectionContext* cntx) {
+  auto opts = ParseReadArgsOrReply(args, read_group, cntx);
+  if (!opts) {
+    return;
+  }
+
+  vector<vector<GroupConsumerPair>> res_pairs(shard_set->size());
+  auto cb = [&](Transaction* t, EngineShard* shard) {
+    if (!opts->read_group) {
+      return OpStatus::OK;
+    }
+    auto sid = shard->shard_id();
+    auto s_args = t->GetShardArgs(sid);
+    GroupConsumerPairOpts gc_opts = {opts->group_name, opts->consumer_name};
+
+    auto items_res = OpGetGroupConsumerPairs(s_args, t->GetOpArgs(shard), gc_opts);
+    if (items_res) {
+      res_pairs[sid] = items_res.value();
+    }
+    return OpStatus::OK;
+  };
+  cntx->transaction->Schedule();
+  cntx->transaction->Execute(std::move(cb), false);
+
+  for (size_t i = 0; i < shard_set->size(); i++) {
+    auto s_item = res_pairs[i];
+    auto s_args = cntx->transaction->GetShardArgs(i);
+    if (!s_item.size()) {
+      continue;
+    }
+    for (size_t j = 0; j < s_args.size(); j++) {
+      string_view key = s_args[j];
+      StreamIDsItem& item = opts->stream_ids.at(key);
+      item.consumer = s_item[j].consumer;
+      item.group = s_item[j].group;
+    }
+  }
+  return XReadGeneric2(args, opts, cntx);
```

Comment:
Probably not the best name for the function.  I see it's being called only here. What does it do? Please add description above the function definition.

Reply:
`XReadGeneric2` is a generic helper function for `XReadGeneric`. I named it like the way `OpSetId2` is. Will add comments :)